### PR TITLE
[SQL] Make date/time functions more consistent with other database systems.

### PR DESCRIPTION
**Pull request by rxin**
This pull request fixes some of the problems in #6981.
- Added date functions to `__all__` so they get exposed
- Rename day_of_month -> dayofmonth
- Rename day_in_year -> dayofyear
- Rename week_of_year -> weekofyear
- Removed "day" from Scala/Python API since it is ambiguous. Only leaving the alias in SQL.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala [hunk 1]
```diff
@@ -1748,182 +1748,6 @@ object functions {
    */
   def length(columnName: String): Column = length(Column(columnName))
 
-  //////////////////////////////////////////////////////////////////////////////////////////////
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: note that this previously cut right into the middle of string functions so I moved them

**Comment by rxin** [comment]
cc @tarekauel and @davies

**Comment by davies** [comment]
LGTM

**Comment by tarekauel** [comment]
@rxin Could you do this little fix as well?
https://github.com/apache/spark/pull/7505/files

Why do we switch from day_of_month to dayofmonth? Most SQL implementations use underscores:
[MySQL](https://dev.mysql.com/doc/refman/5.0/en/func-op-summary-ref.html) [SAP HANA](http://help.sap.com/saphelp_hanaplatform/helpdata/en/20/9f228975191014baed94f1b69693ae/content.htm?frameset=/en/20/9ddefe75191014ac249bf78ba2a1e9/frameset.htm&current_toc=/en/2e/1ef8b4f4554739959886e55d4c127b/plain.htm&node_id=91&show_children=false) [Oracle](http://docs.oracle.com/cd/B19306_01/server.102/b14200/functions001.htm#i88891)
I would prefer underscores, because they improve the readability, if you write all SQL stuff in caps, like:
`SELECT name, age, DAY_OF_MONTH(birthday) AS birthday FROM people WHERE age > 15` compared to `SELECT name, age, DAYOFMONTH(birthday) AS birthday FROM people WHERE age > 15`
I'm not a Python pro, but I thought that underscores are 'pythonic', aren't they?

**Comment by rxin** [comment]
Both MySQL and HANA use dayofmonth, without the underscore?

**Comment by rxin** [comment]
I've merged this.

**Pull request closed without merging**